dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java

```
@@ -66,87 +67,117 @@ public final class ArgumentTypesResolver {
+                    ImmutableOptionalType.builder()
+                            .isPresentMethodName("isPresent")
+                            .valueGetMethodName("getAsInt")
+                            .innerType(getPrimitiveType(context.getTypeMirror(Integer.class))
+                                    .get())
+                            .build()));
+        }
+
+        if (context.isSameTypes(typeMirror, OptionalLong.class)) {
+            return Optional.of(ArgumentTypes.optional(
+                    typeName,
+                    ImmutableOptionalType.builder()
+                            .isPresentMethodName("isPresent")
+                            .valueGetMethodName("getAsLong")
+                            .innerType(getPrimitiveType(context.getTypeMirror(Long.class))
+                                    .get())
+                            .build()));
         }
 
-        return context.getGenericInnerType(Optional.class, typeName)
-                .flatMap(innerType -> getArgumentTypeImpl(paramContext, innerType)
-                        .map(argumentType -> ImmutableOptionalType.builder()
+        if (context.isSameTypes(typeMirror, OptionalDouble.class)) {
+            return Optional.of(ArgumentTypes.optional(
+                    typeName,
+                    ImmutableOptionalType.builder()
+                            .isPresentMethodName("isPresent")
+                            .valueGetMethodName("getAsDouble")
+                            .innerType(getPrimitiveType(context.getTypeMirror(Double.class))
+                                    .get())
+                            .build()));
+        }
+
+        return context.getGenericInnerType(Optional.class, typeMirror)
+                .map(innerType -> getPrimitiveType(innerType)
+                        .or(() -> getAliasType(innerType))
+                        .orElseGet(() -> getCustomType(typeMirror)))
+                .map(innerType -> ArgumentTypes.optional(
```

After this PR, we no longer support the following types:

- `Optional<Optional<T>>`
- `Optional<List<T>>`

We almost certainly never intended to support these types because servers cannot distinguish between `empty()` and `of(empty())`. It is trivial for clients to flatten these values into `Optional<T>` and `List<T>`, respectively.

---

Makes sense. We should also fail loudly on compile when we encounter these cases (where users do not provide an explicit encoder).